programs/lz4io.c

```diff
@@ -395,8 +405,83 @@ typedef struct {
     void*  dstBuffer;
     size_t dstBufferSize;
     LZ4F_compressionContext_t ctx;
+    LZ4F_CDict* cdict;
 } cRess_t;
 
+static void* LZ4IO_createDict(const char* dictFilename, size_t *dictSize) {
+    size_t readSize;
+    size_t dictEnd = 0;
+    size_t dictLen = 0;
+    size_t dictStart;
+    size_t circularBufSize = LZ4_MAX_DICT_SIZE;
+    void* circularBuf;
+    void* dictBuf;
+    FILE* dictFile;
+
+    if (!dictFilename) EXM_THROW(25, "Dictionary error : no filename provided");
+
+    circularBuf = malloc(circularBufSize);
+    if (!circularBuf) EXM_THROW(25, "Allocation error : not enough memory");
+
+    dictFile = LZ4IO_openSrcFile(dictFilename);
+    if (!dictFile) EXM_THROW(25, "Dictionary error : could not open dictionary file");
+
+    if (strcmp(dictFilename, stdinmark)) {
+        UTIL_fseek(dictFile, -LZ4_MAX_DICT_SIZE, SEEK_END);
+    }
+
+    do {
+        readSize = fread((char*)circularBuf + dictEnd, (size_t)1, circularBufSize - dictEnd, dictFile);
+        dictEnd = (dictEnd + readSize) % LZ4_MAX_DICT_SIZE;
+        dictLen += readSize;
+    } while (readSize>0);
+
+    if (dictLen > LZ4_MAX_DICT_SIZE) {
+        dictLen = LZ4_MAX_DICT_SIZE;
+    }
+
+    *dictSize = dictLen;
+
+    dictStart = (circularBufSize + dictEnd - dictLen) % circularBufSize;
+
+    if (dictStart == 0) {
+        // We're in the simple case where the dict starts at the beginning of our circular buffer.
+        dictBuf = realloc(circularBuf, dictLen ? dictLen : 1);
```

Comment:
We don't need this `realloc`, we can just say

```c
dictBuf = circularBuf;
circularBuf = NULL;
```

Reply:
The `realloc` is to shrink the buffer if the dict is smaller than 64KB.

Reply:
Maybe that's not valuable, though?

Reply:
We free the buffer very soon after, when we create the CDict right?

In any case, I'm not too worried by over allocating by up to 64 KB.

Reply:
Ah, that's a good point. We do on the compress path, but not on the decompress path. Ok, I'll axe it.